equinox/_ad.py

```diff
@@ -940,3 +940,82 @@ def filter_custom_vjp(fn):
 
     filter_custom_jvp.__doc__ = _filter_custom_jvp_doc
     filter_custom_vjp.__doc__ = _filter_custom_vjp_doc
+
+
+def filter_checkpoint(
+    fun: Callable[_P, _T] = sentinel,
+    *,
+    prevent_cse: bool = True,
+    policy: Optional[Callable[..., bool]] = None,
+) -> Callable[_P, _T]:
+    """Filtered version of `jax.checkpoint`.
+
+    Gradient checkpointing is a technique for reducing memory usage during
+    backpropagation, especially when used with reverse mode automatic differentiation
+    (e.g., `jax.grad` or `equinox.filter_grad`).
+
+    **Arguments:**
+
+    - `fun`: The function to be checkpointed. Will be called as `fun(*args, **kwargs)`.
+        Can return an arbitrary PyTree.
+    - `prevent_cse`: If `True` (the default), then JAX will not perform common
+    subexpression elimination. Please see the documentation for `jax.checkpoint
+    ` for more details.
+    - `policy`: Callable for controlling which intermediate values should be
+    rematerialized. It should be one of the attributes of `jax.checkpoint_policies`.
+    """
+
+    if fun is sentinel:
+        return ft.partial(  # pyright: ignore
+            filter_checkpoint,
+            prevent_cse=prevent_cse,
+            policy=policy,
+        )
+
+    return module_update_wrapper(
+        _CheckpointWrapper(fun, prevent_cse=prevent_cse, policy=policy), fun
+    )
+
+
+class _CheckpointWrapper(Module):
+    _fun: Callable
+    _prevent_cse: bool
+    _policy: Optional[Callable[..., bool]]
+
+    def __init__(
+        self,
+        fun: Callable,
+        *,
+        prevent_cse: bool = True,
+        policy: Optional[Callable[..., bool]] = None,
+    ):
+        self._fun = fun
+        self._prevent_cse = prevent_cse
+        self._policy = policy
+
+    @property
+    def __wrapped__(self):
+        return self._fun
+
+    def __call__(self, *args, **kwargs):
+        @ft.partial(
+            jax.checkpoint,  # pyright: ignore
+            prevent_cse=self._prevent_cse,
+            policy=self._policy,
+            static_argnums=(0,),
+        )
+        def fun_checkpoint(_static, _dynamic):
+            _args, _kwargs = combine(_static, _dynamic)
+            out = self._fun(*_args, **_kwargs)
+            _dynamic_out, _static_out = partition(out, _is_struct)
```

Comment:
Is `_is_struct` really the criterion we need here? (On both the input and the output.) I think just `eqx.is_array` would be correct?

Reply:
yeah you're right. will fix. do you like is_array or is_array_like here? i feel like probably is_array is fine since you never need to checkpoint scalars (except with some truly massive vmaps or xmaps or something)

Reply:
I think `is_array` is probably best here. (Equinox generally prefers to distinguish arrays from arraylikes.)

Reply:
ok, so all done then!